source/common/buffer/buffer_impl.cc

```diff
@@ -94,7 +95,43 @@ void OwnedImpl::move(Instance& rhs, uint64_t length) {
 }
 
 int OwnedImpl::read(int fd, uint64_t max_length) {
-  return evbuffer_read(buffer_.get(), fd, max_length);
+  if (max_length == 0) {
+    return 0;
+  }
+  constexpr uint64_t MaxSlices = 2;
+  RawSlice slices[MaxSlices];
+  const uint64_t num_slices = reserve(max_length, slices, MaxSlices);
+  struct iovec iov[num_slices];
+  uint64_t num_slices_to_read = 0;
+  uint64_t num_bytes_to_read = 0;
+  for (; num_slices_to_read < num_slices && num_bytes_to_read < max_length; num_slices_to_read++) {
+    iov[num_slices_to_read].iov_base = slices[num_slices_to_read].mem_;
+    const size_t slice_length = std::min(slices[num_slices_to_read].len_,
+                                         static_cast<size_t>(max_length - num_bytes_to_read));
+    iov[num_slices_to_read].iov_len = slice_length;
+    num_bytes_to_read += slice_length;
+  }
+  ASSERT(num_slices_to_read <= MaxSlices);
+  ASSERT(num_bytes_to_read <= max_length);
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  const ssize_t rc = os_syscalls.readv(fd, iov, static_cast<int>(num_slices_to_read));
+  if (rc < 0) {
+    return rc;
+  }
+  uint64_t num_slices_to_commit = 0;
+  uint64_t bytes_to_commit = rc;
+  ASSERT(bytes_to_commit <= max_length);
+  while (bytes_to_commit != 0) {
+    if (slices[num_slices_to_commit].len_ > bytes_to_commit) {
+      slices[num_slices_to_commit].len_ = bytes_to_commit;
+    }
+    ASSERT(bytes_to_commit >= slices[num_slices_to_commit].len_);
```

Comment:
These 4 lines seem more verbose than needed. Couldn't this all be shrunk down to `slices[num_slices_to_commit].len_ = std::min(slices[num_slices_to_commit].len_, bytes_to_commit);`? The ASSERT seems redundant, because the previous clause guaranteed the condition.

Reply:
Sorry why do we think the ASSERT is redundant? It caught a pretty bad bug previously. Especially with the way the code is now it seems good to double check the min logic is correct and we don't overflow?